scale/scale-codegen/src/main/java/com/strategyobject/substrateclient/scale/codegen/reader/ScaleReaderAnnotatedEnum.java

```diff
@@ -0,0 +1,71 @@
+import com.strategyobject.substrateclient.common.io.Streamer;
+import com.strategyobject.substrateclient.scale.ScaleReader;
+import com.strategyobject.substrateclient.scale.annotation.AutoRegister;
+import com.strategyobject.substrateclient.scale.codegen.ScaleProcessorHelper;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.TypeElement;
+import java.io.IOException;
+import java.io.InputStream;
+
+@RequiredArgsConstructor
+public class ScaleReaderAnnotatedEnum {
+    private static final String READERS_ARG = "readers";
+
+    private final @NonNull TypeElement enumElement;
+
+    public void generateReader(@NonNull ProcessorContext context) throws IOException, ProcessingException {
+        val readerName = ScaleProcessorHelper.getReaderName(enumElement.getSimpleName().toString());
+        val enumType = TypeName.get(enumElement.asType());
+
+        val typeSpecBuilder = TypeSpec.classBuilder(readerName)
+                .addAnnotation(AnnotationSpec.builder(AutoRegister.class)
+                        .addMember("types", "{$L.class}", enumElement.getQualifiedName().toString())
+                        .build())
+                .addModifiers(Modifier.PUBLIC)
+                .addSuperinterface(ParameterizedTypeName.get(ClassName.get(ScaleReader.class), enumType))
+                .addMethod(generateReadMethod(enumType));
+
+        JavaFile.builder(
+                context.getPackageName(enumElement),
+                typeSpecBuilder.build()
+        ).build().writeTo(context.getFiler());
+    }
+
+    private MethodSpec generateReadMethod(TypeName enumType) {
+        val methodSpec = MethodSpec.methodBuilder("read")
+                .addAnnotation(Override.class)
+                .addModifiers(Modifier.PUBLIC)
+                .returns(enumType)
+                .addParameter(InputStream.class, "stream")
+                .addParameter(ArrayTypeName.of(
+                                ParameterizedTypeName.get(
+                                        ClassName.get(ScaleReader.class),
+                                        WildcardTypeName.subtypeOf(Object.class))),
+                        READERS_ARG)
+                .varargs(true)
+                .addException(IOException.class);
+
+        addValidationRules(methodSpec);
+        addMethodBody(methodSpec);
+        return methodSpec.build();
+    }
+
+    private void addValidationRules(MethodSpec.Builder methodSpec) {
+        methodSpec
+                .addStatement("if (stream == null) throw new IllegalArgumentException(\"stream is null\")")
+                .addStatement("if (readers != null && readers.length > 0) throw new IllegalArgumentException()");
+    }
+
+    private void addMethodBody(MethodSpec.Builder methodSpec) {
+        methodSpec.addStatement("return $T.values()[$T.readByte(stream)]", enumElement, Streamer.class);
```

Comment:
This method will allocate array every time. I would prefer to iterate over enum constants and generate `switch ` statement or something like this. I think there also should be a check of byte read from scale doesn't exceed number of available constants.

Reply:
I saved `values` to a field, so that it is not cloned on each read and added a helper method with exception handling.